redux/redux.d.ts

```diff
@@ -1,49 +1,84 @@
-// Type definitions for Redux v1.0.0
+// Type definitions for Redux v3.0.0
 // Project: https://github.com/rackt/redux
 // Definitions by: William Buchwalter <https://github.com/wbuchwalter/>, Vincent Prouillet <https://github.com/Keats/>
 // Definitions: https://github.com/borisyankov/DefinitelyTyped
 
 declare module Redux {
+    // Minimalist Action definition
+    interface Action {
+        type: string;
+    }
+    
+    // FSA-compliant action (from redux-actions.d.ts)
+    // See: https://github.com/acdlite/flux-standard-action
+    interface FluxStandardAction<P,M> extends Action {
+        payload?: P
+        error?: boolean
+        meta?: M
+    }
 
+    // Not used...?
     interface ActionCreator extends Function {
-        (...args: any[]): any;
+        (...args: any[]): Action;
     }
 
-    interface Reducer extends Function {
-        (state: any, action: any): any;
+    // Also not used here...
+    interface StoreMethods<T,A extends Action> {
+        dispatch: Dispatch;
+        getState(): T;
     }
 
+    // Reducer is a function that takes a state of type T and an
+    // action and returns a new state, also of type T
+    interface Reducer<T> extends Function {
+        (state: T, action: any): T;
+    }
+
+    // Dispatch is a function that takes an action of type A and
+    // returns an instance of that same action (still of type A)
     interface Dispatch extends Function {
-        (action: any): any;
+        <A extends Action>(action: A): A;
     }
 
-    interface StoreMethods {
+    // MiddlewareArg is an instance of middleware that provides a
+    // dispatch method and a getState method.  The type parameter T is
+    // the type of the state.  No type parameter is provided on
+    // Dispatch because the middleware generally needs to be able to
+    // handle any type of action.
+    interface MiddlewareArg<T> {
         dispatch: Dispatch;
-        getState(): any;
+        getState: () => T;
     }
 
-
-    interface MiddlewareArg {
-        dispatch: Dispatch;
-        getState: Function;
+    // Middleware is constructed from an instance of MiddlewareArg<T> where
+    // T is the type of the state.
+    interface Middleware<T> extends Function {
+        (obj: MiddlewareArg<T>): Function;
     }
 
-    interface Middleware extends Function {
-        (obj: MiddlewareArg): Function;
+    // Store manages the state of the system.  The state of the system
+    // should be of type T.
+    class Store<T> {
+        getReducer(): Reducer<T>;
+        replaceReducer(nextReducer: Reducer<T>): void;
+        dispatch<A extends Action>(action: A): A;
```

Comment:
store.dispatch(action) can return, for example, a Promise, if some middleware process this action